Android 9.0 多线程使用相同目录问题

Android9.0机型上出现 闪退  报上放的错

04-23 19:18:04.657 17580-17580/com.ibgoing E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ibgoing, PID: 17580
 java.lang.RuntimeException: Using WebView from more than one process at once with the same data directory is not supported. https://crbug.com/558377
 at com.android.webview.chromium.WebViewChromiumAwInit.startChromiumLocked(WebViewChromiumAwInit.java:63)
     at 
at com.android.webview.chromium.WebViewChromiumAwInitForP.startChromiumLocked(WebViewChromiumAwInitForP.java:3)
  at com.android.webview.chromium.WebViewChromiumAwInit$3.run(WebViewChromiumAwInit.java:3)  at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
 at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6863)
        at java.lang.reflect.Method.invoke(Native Method)
 at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)

我也遇到，也是搜到的楼上的解决方案：
```text
1、异常日志:Caused by: java.lang.RuntimeException: Using WebView from more than one process at once with the same data directory is not supported. https://crbug.com/558377

2、异常出现情景:因为Android P行为变更，不可多进程使用同一个目录webView，需要为不同进程webView设置不同目录
3、解决方式：重写项目Application，然后为其它进程webView设置目录
————————————————
版权声明：本文为CSDN博主「erweimaerweima」的原创文章，遵循 CC 4.0 BY-SA 版权协议，转载请附上原文出处链接及本声明。
原文链接：https://blog.csdn.net/lvshuchangyin/article/details/89446629
```

--- a/agentweb-core/src/main/java/com/just/agentweb/AbsAgentWebSettings.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/AbsAgentWebSettings.java
@@ -17,6 +17,7 @@
 package com.just.agentweb;
 
 
+import android.content.Context;
 import android.os.Build;
 import android.view.View;
 import android.webkit.DownloadListener;
@@ -125,6 +126,15 @@ public abstract class AbsAgentWebSettings implements IAgentWebSettings, WebListe
 				.concat(USERAGENT_UC)
 		);
 		LogUtils.i(TAG, "UserAgentString : " + mWebSettings.getUserAgentString());
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+			// 安卓9.0后不允许多进程使用同一个数据目录，需设置前缀来区分
+			// 参阅 https://blog.csdn.net/lvshuchangyin/article/details/89446629
+			Context context = webView.getContext();
+			String processName = ProcessUtils.getCurrentProcessName(context);
+			if (!context.getApplicationContext().getPackageName().equals(processName)) {
+				WebView.setDataDirectorySuffix(processName);
+			}
+		}
 	}
 
 	@Override

--- a/agentweb-core/src/main/java/com/just/agentweb/AgentWebSettingsImpl.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/AgentWebSettingsImpl.java
@@ -17,6 +17,8 @@
 package com.just.agentweb;
 
 import android.app.Activity;
+import android.content.Context;
+import android.content.ContextWrapper;
 import android.webkit.DownloadListener;
 import android.webkit.WebView;
 
@@ -26,17 +28,36 @@ import android.webkit.WebView;
  * @since 1.0.0
  */
 public class AgentWebSettingsImpl extends AbsAgentWebSettings {
-	private AgentWeb mAgentWeb;
+    private AgentWeb mAgentWeb;
 
-	@Override
-	protected void bindAgentWebSupport(AgentWeb agentWeb) {
-		this.mAgentWeb = agentWeb;
-	}
+    @Override
+    protected void bindAgentWebSupport(AgentWeb agentWeb) {
+        this.mAgentWeb = agentWeb;
+    }
 
-	@Override
-	public WebListenerManager setDownloader(WebView webView, DownloadListener downloadListener) {
-		return super.setDownloader(webView, downloadListener == null ?
-				DefaultDownloadImpl.create((Activity) webView.getContext(), webView, mAgentWeb.getPermissionInterceptor()) :
-				downloadListener);
-	}
+    @Override
+    public WebListenerManager setDownloader(WebView webView, DownloadListener downloadListener) {
+        // Fix Android 5.1 crashing:
+        // ClassCastException: android.app.ContextImpl cannot be cast to android.app.Activity
+        if (downloadListener == null) {
+            Activity activity = getActivityByContext(webView.getContext());
+            downloadListener = DefaultDownloadImpl.create(activity, webView, mAgentWeb.getPermissionInterceptor());
+        }
+        return super.setDownloader(webView, downloadListener);
+    }
+
+    /**
+     * Copy from com.blankj.utilcode.util.ActivityUtils#getActivityByView
+     */
+    private Activity getActivityByContext(Context context) {
+        if (context instanceof Activity) return (Activity) context;
+        while (context instanceof ContextWrapper) {
+            if (context instanceof Activity) {
+                return (Activity) context;
+            }
+            context = ((ContextWrapper) context).getBaseContext();
+        }
+        return null;
+    }
+
 }

--- a/agentweb-core/src/main/java/com/just/agentweb/AgentWebUtils.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/AgentWebUtils.java
@@ -440,7 +440,7 @@ public class AgentWebUtils {
 
 		try {
 
-			clearWebViewAllCache(context, new WebView(context.getApplicationContext()));
+			clearWebViewAllCache(context, new LollipopFixedWebView(context.getApplicationContext()));
 		} catch (Exception e) {
 			e.printStackTrace();
 		}

--- a/agentweb-core/src/main/java/com/just/agentweb/AgentWebView.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/AgentWebView.java
@@ -48,7 +48,7 @@ import java.util.Map;
  * @author cenxiaozhong
  * @since 1.0.0
  */
-public class AgentWebView extends WebView {
+public class AgentWebView extends LollipopFixedWebView {
     private static final String TAG = AgentWebView.class.getSimpleName();
     private Map<String, JsCallJava> mJsCallJavas;
     private Map<String, String> mInjectJavaScripts;

--- a/agentweb-core/src/main/java/com/just/agentweb/DefaultWebCreator.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/DefaultWebCreator.java
@@ -229,7 +229,7 @@ public class DefaultWebCreator implements WebCreator {
             mWebView = new AgentWebView(mActivity);
             AgentWebConfig.WEBVIEW_TYPE = AgentWebConfig.WEBVIEW_AGENTWEB_SAFE_TYPE;
         } else {
-            mWebView = new WebView(mActivity);
+            mWebView = new LollipopFixedWebView(mActivity);
             AgentWebConfig.WEBVIEW_TYPE = AgentWebConfig.WEBVIEW_DEFAULT_TYPE;
         }
         return mWebView;

--- a/agentweb-core/src/main/java/com/just/agentweb/LollipopFixedWebView.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/LollipopFixedWebView.java
@@ -1,0 +1,45 @@
+package com.just.agentweb;
+
+import android.annotation.TargetApi;
+import android.content.Context;
+import android.content.res.Configuration;
+import android.os.Build;
+import android.util.AttributeSet;
+import android.webkit.WebView;
+
+/**
+ * 修复 Android 5.0 & 5.1 打开 WebView 闪退问题：
+ * 参阅 https://stackoverflow.com/questions/41025200/android-view-inflateexception-error-inflating-class-android-webkit-webview
+ */
+@SuppressWarnings("unused")
+public class LollipopFixedWebView extends WebView {
+    public LollipopFixedWebView(Context context) {
+        super(getFixedContext(context));
+    }
+
+    public LollipopFixedWebView(Context context, AttributeSet attrs) {
+        super(getFixedContext(context), attrs);
+    }
+
+    public LollipopFixedWebView(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(getFixedContext(context), attrs, defStyleAttr);
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public LollipopFixedWebView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
+        super(getFixedContext(context), attrs, defStyleAttr, defStyleRes);
+    }
+
+    public LollipopFixedWebView(Context context, AttributeSet attrs, int defStyleAttr, boolean privateBrowsing) {
+        super(getFixedContext(context), attrs, defStyleAttr, privateBrowsing);
+    }
+
+    public static Context getFixedContext(Context context) {
+        if (Build.VERSION.SDK_INT >= 21 && Build.VERSION.SDK_INT < 23) {
+            // Avoid crashing on Android 5 and 6 (API level 21 to 23)
+            return context.createConfigurationContext(new Configuration());
+        }
+        return context;
+    }
+
+}

--- a/agentweb-core/src/main/java/com/just/agentweb/ProcessUtils.java
+++ b/agentweb-core/src/main/java/com/just/agentweb/ProcessUtils.java
@@ -1,0 +1,78 @@
+package com.just.agentweb;
+
+import android.app.ActivityManager;
+import android.app.Application;
+import android.content.Context;
+import android.text.TextUtils;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.List;
+
+/**
+ * Adapted from com.blankj.utilcode.util.ProcessUtils#getCurrentProcessName
+ */
+class ProcessUtils {
+
+    static String getCurrentProcessName(Context context) {
+        String name = getCurrentProcessNameByFile();
+        if (!TextUtils.isEmpty(name)) return name;
+        name = getCurrentProcessNameByAms(context);
+        if (!TextUtils.isEmpty(name)) return name;
+        name = getCurrentProcessNameByReflect(context);
+        return name;
+    }
+
+    private static String getCurrentProcessNameByFile() {
+        try {
+            File file = new File("/proc/" + android.os.Process.myPid() + "/" + "cmdline");
+            BufferedReader mBufferedReader = new BufferedReader(new FileReader(file));
+            String processName = mBufferedReader.readLine().trim();
+            mBufferedReader.close();
+            return processName;
+        } catch (Exception e) {
+            e.printStackTrace();
+            return "";
+        }
+    }
+
+    private static String getCurrentProcessNameByAms(Context context) {
+        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
+        if (am == null) return "";
+        List<ActivityManager.RunningAppProcessInfo> info = am.getRunningAppProcesses();
+        if (info == null || info.size() == 0) return "";
+        int pid = android.os.Process.myPid();
+        for (ActivityManager.RunningAppProcessInfo aInfo : info) {
+            if (aInfo.pid == pid) {
+                if (aInfo.processName != null) {
+                    return aInfo.processName;
+                }
+            }
+        }
+        return "";
+    }
+
+    private static String getCurrentProcessNameByReflect(Context context) {
+        String processName = "";
+        try {
+            Application app = (Application) context.getApplicationContext();
+            Field loadedApkField = app.getClass().getField("mLoadedApk");
+            loadedApkField.setAccessible(true);
+            Object loadedApk = loadedApkField.get(app);
+
+            Field activityThreadField = loadedApk.getClass().getDeclaredField("mActivityThread");
+            activityThreadField.setAccessible(true);
+            Object activityThread = activityThreadField.get(loadedApk);
+
+            Method getProcessName = activityThread.getClass().getDeclaredMethod("getProcessName");
+            processName = (String) getProcessName.invoke(activityThread);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return processName;
+    }
+
+}

